File: ports/csharp/src/core/global/BcsSignal.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace BCS.Core.Global {
    public class BcsSignal<T> {
        private readonly List<Action<T>> _listeners = new List<Action<T>>();
        private readonly object _lock = new object();

        public void Connect(Action<T> callback) {
            lock (_lock) {
                _listeners.Add(callback);
            }
        }

        public void Emit(T payload) {
            List<Action<T>> listenersCopy;
            lock (_lock) {
                listenersCopy = new List<Action<T>>(_listeners);
            }

            foreach (var listener in listenersCopy) {
                Task.Run(() => listener(payload));
            }
        }
    }
}

File: ports/java/src/core/global/BcsSignal.java
package bcs.core.global;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.function.Consumer;

public class BcsSignal<T> {
    private final List<Consumer<T>> listeners;
    private static final ExecutorService executor = Executors.newCachedThreadPool();

    public BcsSignal() {
        this.listeners = new CopyOnWriteArrayList<>();
    }

    public void connect(Consumer<T> callback) {
        this.listeners.add(callback);
    }

    public void emit(T payload) {
        for (Consumer<T> listener : this.listeners) {
            executor.submit(() -> listener.accept(payload));
        }
    }
}
